apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+                    }
+                    var msgs = pull.ResponseStream.Current.ReceivedMessages;
+                    var ackIds = new HashSet<string>(msgs.Select(x => x.AckId));
+                    var allMsgsHandledCts = CancellationTokenSource.CreateLinkedTokenSource(_hardStopCts.Token);
+                    _registerTaskFn(_taskHelper.Run(() => AutoExtend(ackIds, allMsgsHandledCts.Token)));
+                    foreach (var msg in msgs)
+                    {
+                        if (cts.IsCancellationRequested)
+                        {
+                            break;
+                        }
+                        // Returned Task completes once flow-control is satisfied.
+                        await _taskHelper.ConfigureAwait(_flow.Process(msg.CalculateSize(), async () =>
+                        {
+                            if (cts.IsCancellationRequested)
+                            {
+                                return;
+                            }
+                            lock (_qLock)
+                            {
+                                _unAckedMsgCount += 1;
+                            }
+                            var reply = await _taskHelper.ConfigureAwaitHideErrors(_handlerAsync(msg.Message, _hardStopCts.Token), Reply.Nack);
+                            TaskCompletionSource<int> qEvent = null;
+                            lock (_qLock)
+                            {
+                                if (reply == Reply.Ack)
+                                {
+                                    ackIds.Remove(msg.AckId);
+                                    _ackQueue.Enqueue(msg.AckId);
+                                    if (ackIds.Count == 0)
+                                    {
+                                        allMsgsHandledCts.Cancel();
+                                    }
+                                    qEvent = _qEvent;
+                                }
+                                else
+                                {
+                                    _unAckedMsgCount -= 1;
+                                }
+                            }
+                            if (qEvent != null)
+                            {
+                                qEvent.TrySetResult(0);
+                            }
+                        }));
+                    }
+                }
+            }
+
+            private async Task AutoExtend(HashSet<string> ackIds, CancellationToken allMsgsHandledToken)
+            {
+                // Moves ackIds into the extend-queue as required for auto lease extension
+                while (true)
+                {
+                    await _taskHelper.ConfigureAwait(_scheduler.Delay(_autoExtendInterval, allMsgsHandledToken));
+                    TaskCompletionSource<int> qEvent;
+                    lock (_qLock)
+                    {
+                        _extendQueue.EnqueueAll(ackIds);
```

Comment:
Okay, so this takes a snapshot of the ack IDs that hadn't been processed when the timer went off... which may still be processed, but that's okay?